### ISSUE TITLE
Pods sometimes get assigned the same IP

### ISSUE THREAD

**Issue by narasi** [kubernetes/kubernetes#51322]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**: 
I was using Kubernetes to setup a bunch of demo environments, and noticed that sometimes, it appeared as if the loadbalancer was treating the pod of another deployment as its own pod. After some debugging I found out that when the pods are brought up, they got assigned the same IP.

**What you expected to happen**:
Kubernetes must poll for IPs that are already taken before assigning an IP to a pod.

**How to reproduce it (as minimally and precisely as possible)**:
I actually do not know how to reproduce the problem. We did not have any infrastructure level events, except that I was handling a bunch of deployments at that time.

**Anything else we need to know?**:
Two things that are to be noted: 

1. I noticed this problem occured over. Three deployments were affected, where on different instances, two pods belonging to different deployments shared the same IP address. 

```
$ kubectl get pods -o wide
NAME                                 READY     STATUS    RESTARTS   AGE       IP            NODE
k8sacc-drupaldemo-2612891940-74qr9   1/1       Running   0          5d        10.240.0.25   k8s-agent-82f8582-1
k8sacc-drupaldev-1306823299-gvsjf    1/1       Running   0          3d        10.240.0.26   k8s-agent-82f8582-1
k8sacc-drupalqa-3650587400-tkrcr     1/1       Running   0          2h        10.240.0.25   k8s-agent-82f8582-2
k8sacc-drupalux-1958487503-05709     1/1       Running   0          26d       10.241.0.6    k8s-agent-82f8582-0
k8sacc-drupalux-1958487503-95xx3     1/1       Running   0          26d       10.240.0.19   k8s-agent-82f8582-2
```

The second time round:
```
$ kubectl get pods -o wide
NAME                                 READY     STATUS        RESTARTS   AGE       IP            NODE
k8sacc-drupaldemo-2122692172-1w04g   1/1       Running       0          1h        10.240.0.26   k8s-agent-82f8582-2
k8sacc-drupaldev-1306823299-gvsjf    1/1       Running       0          3d        10.240.0.26   k8s-agent-82f8582-1
k8sacc-drupalqa-3650587400-tkrcr     1/1       Running       0          5h        10.240.0.25   k8s-agent-82f8582-2
k8sacc-drupalux-1958487503-05709     1/1       Running       0          27d       10.241.0.6    k8s-agent-82f8582-0
k8sacc-drupalux-1958487503-95xx3     1/1       Running       0          27d       10.240.0.19   k8s-agent-82f8582-2
```

2. This raised questions of what about the pods within the kube-system namespace. I noticed that the dashboard was having a `CrashLoopBackOff` and had lots of restarts just because it got assigned an IP that was already taken, I presume.

```
$ kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                                           READY     STATUS             RESTARTS   AGE       IP             NODE
default       k8sacc-drupaldemo-2122692172-dh3c7             1/1       Running            0          15h       10.240.0.30    k8s-agent-82f8582-1
default       k8sacc-drupaldev-1306823299-gvsjf              1/1       Running            0          4d        10.240.0.26    k8s-agent-82f8582-1
default       k8sacc-drupalqa-3650587400-tkrcr               1/1       Running            0          21h       10.240.0.25    k8s-agent-82f8582-2
default       k8sacc-drupalux-1958487503-05709               1/1       Running            0          27d       10.241.0.6     k8s-agent-82f8582-0
default       k8sacc-drupalux-1958487503-95xx3               1/1       Running            0          27d       10.240.0.19    k8s-agent-82f8582-2
kube-system   heapster-v1.2.0-1516787090-x21g1               2/2       Running            0          45d       10.241.0.4     k8s-agent-82f8582-0
kube-system   kube-addon-manager-k8s-master-82f8582-0        1/1       Running            22         45d       10.241.255.5   k8s-master-82f8582-0
kube-system   kube-apiserver-k8s-master-82f8582-0            1/1       Running            194        45d       10.241.255.5   k8s-master-82f8582-0
kube-system   kube-controller-manager-k8s-master-82f8582-0   1/1       Running            34         45d       10.241.255.5   k8s-master-82f8582-0
kube-system   kube-dns-v19-2jf6h                             2/3       Running            35228      45d       10.241.0.3     k8s-agent-82f8582-0
kube-system   kube-dns-v19-qdw2c                             1/3       CrashLoopBackOff   35235      45d       10.241.0.2     k8s-agent-82f8582-0
kube-system   kube-proxy-6nm6g                               1/1       Running            0          45d       10.241.0.6     k8s-agent-82f8582-0
kube-system   kube-proxy-hgb9h                               1/1       Running            0          45d       10.241.255.5   k8s-master-82f8582-0
kube-system   kube-proxy-r915m                               1/1       Running            0          45d       10.241.0.5     k8s-agent-82f8582-1
kube-system   kube-proxy-w89l8                               1/1       Running            0          45d       10.241.0.4     k8s-agent-82f8582-2
kube-system   kube-scheduler-k8s-master-82f8582-0            1/1       Running            31         45d       10.241.255.5   k8s-master-82f8582-0
kube-system   kubernetes-dashboard-3868706175-3t6f5          0/1       CrashLoopBackOff   12568      45d       10.241.0.5     k8s-agent-82f8582-0
```

> Therefore, this seems to be one of the things where kubernetes, if it thinks the environment is misconfigured, it must fail instead of trying to make it work.

^ lol I guess I was trying to say two things and merged the sentences. 
Kubernetes has issues with the environment and tries to fix it by plain restart so many times. Perhaps after a point it must delete the pod and try to bring it back up with the added checks instead of trying infinitely. I get that this is a risky strategy for the `kube-system` pods, but that is just too many restarts for a problem that was apparently fixed by a delete 😅 

**Environment**:
- Kubernetes version (use `kubectl version`): 
```$ kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.6", GitCommit:"7fa1c1756d8bc963f1a389f4a6937dc71f08ada2", GitTreeState:"clean", BuildDate:"2017-06-16T18:21:54Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.6", GitCommit:"7fa1c1756d8bc963f1a389f4a6937dc71f08ada2", GitTreeState:"clean", BuildDate:"2017-06-16T18:21:54Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration**: Azure
- OS (e.g. from /etc/os-release): Ubuntu 16.04.2 LTS
- Kernel (e.g. `uname -a`): 
```Linux k8s-master-82F8582-0 4.4.0-83-generic #106-Ubuntu SMP Mon Jun 26 17:54:43 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux```
- Install tools: Azure ARM template
- Others:

**Edited for more clarity.

**Comment by cmluciano** [comment]
@narasi Can you briefly describe the performance or setup process that you were going through?

You mentioned that you had a high number of deployments. Do you have a rough guess on the number of deployments you were at when this started to occur.

 Can you also share your cluster size?

**Comment by cmluciano** [comment]
cc @kubernetes/sig-azure-misc

**Comment by narasi** [comment]
@cmluciano : 
So this cluster was setup using the Azure Resource Template, and I had been running it for our acceptance and demo purposes for quite some time. 

Right before the problem surfaced, I was making two deployments. Up until this point, we did not see mixed requests yet. 

Once the problem surfaced, I started fiddling around with the pods to see why exactly they were serving mixed requests. It was a few hours before it hit me. Fiddling around : Attaching a bash session to tail logs, just re-verify the run time setup etc.

P.S.: I handled 4 different deployments today on the same cluster, and did not observe the issue recur. It is worrying though.

**Comment by feiskyer** [comment]
```
$ kubectl get pods -o wide
NAME                                 READY     STATUS    RESTARTS   AGE       IP            NODE
k8sacc-drupaldemo-2612891940-74qr9   1/1       Running   0          5d        10.240.0.25   k8s-agent-82f8582-1
k8sacc-drupalqa-3650587400-tkrcr     1/1       Running   0          2h        10.240.0.25   k8s-agent-82f8582-2
```

```
$ kubectl get pods -o wide
NAME                                 READY     STATUS        RESTARTS   AGE       IP            NODE
k8sacc-drupaldemo-2122692172-1w04g   1/1       Running       0          1h        10.240.0.26   k8s-agent-82f8582-2
k8sacc-drupaldev-1306823299-gvsjf    1/1       Running       0          3d        10.240.0.26   k8s-agent-82f8582-1
```

@narasi From the first and second round, pods running on different nodes got same IPs, which is not expected. So I think you have miss configured the CNI network plugin. Each node should receive a different subnet. 

May I ask which network plugin are you using? And could you check whether the subnet are configured differently on different nodes?

**Comment by narasi** [comment]
Hi @feiskyer! Thanks for looking into this.

> May I ask which network plugin are you using?
We are using `kubenet`. 

> And could you check whether the subnet are configured differently on different nodes?
We did modify some of the `kubelet` parameters in order to ensure that the subnet is compatible with the subnet of some external services, like databases and logging servers. 
Let me paste the kubelet arguments here

```
[Unit]
Description=Kubelet
Requires=docker.service
After=docker.service

[Service]
Restart=always
EnvironmentFile=/etc/default/kubelet
SuccessExitStatus=143
ExecStartPre=/bin/bash /opt/azure/containers/kubelet.sh
ExecStartPre=/bin/mkdir -p /var/lib/kubelet
ExecStartPre=/bin/bash -c "if [ $(mount | grep \"/var/lib/kubelet\" | wc -l) -le 0 ] ; then /bin/mount --bind /var/lib/kubelet /var/lib/kubelet ; fi"
ExecStartPre=/bin/mount --make-shared /var/lib/kubelet
ExecStartPre=-/sbin/ebtables -t nat --list
ExecStartPre=-/sbin/iptables -t nat --list
ExecStart=/usr/bin/docker run \
  --net=host \
  --pid=host \
  --privileged \
  --rm \
  --volume=/dev:/dev \
  --volume=/sys:/sys:ro \
  --volume=/var/run:/var/run:rw \
  --volume=/var/lib/docker/:/var/lib/docker:rw \
  --volume=/var/lib/kubelet/:/var/lib/kubelet:shared \
  --volume=/var/log:/var/log:rw \
  --volume=/etc/kubernetes/:/etc/kubernetes:ro \
  --volume=/srv/kubernetes/:/srv/kubernetes:ro $DOCKER_OPTS \
    gcrio.azureedge.net/google_containers/hyperkube-amd64:v1.6.6 \
      /hyperkube kubelet \
        --kubeconfig=/var/lib/kubelet/kubeconfig \
        --require-kubeconfig \
        --pod-infra-container-image="gcrio.azureedge.net/google_containers/pause-amd64:3.0" \
        --address=0.0.0.0 \
        --allow-privileged=true \
        --enable-server \
        --enable-debugging-handlers \
        --pod-manifest-path=/etc/kubernetes/manifests \
        --cluster-dns=10.0.0.10 \
        --cluster-domain=cluster.local \
        --register-schedulable=true \
        --node-labels="role=agent" \
        --cloud-provider=azure \
        --cloud-config=/etc/kubernetes/azure.json \
        --azure-container-registry-config=/etc/kubernetes/azure.json \
        --hairpin-mode=promiscuous-bridge \
        --network-plugin=kubenet \
        --pod-cidr=10.241.0.0/16 \
        --non-masquerade-cidr=10.241.0.0/16 \
        --v=2 --feature-gates=Accelerators=true

[Install]
WantedBy=multi-user.target
```

I hope this would answer your questions 😄

**Comment by jdumars** [comment]
@narasi can you confirm this is still a problem in > 1.8.x?  Thanks.

**Comment by Matty9191** [comment]
I'm seeing this  behavior with 1.9.2:

$ kubectl get pods -o wide
NAME      READY     STATUS    RESTARTS   AGE       IP         NODE
log1      1/1       Running   0          4m        **10.1.0.2**   kubworker1
log2      1/1       Running   0          6m        **10.1.0.2**   kubworker3

$ kubectl get nodes
NAME                       STATUS    ROLES     AGE       VERSION
kubworker1   Ready     <none>    21m       v1.9.2
kubworker2   Ready     <none>    21m       v1.9.2
kubworker3   Ready     <none>    21m       v1.9.2

Anyone else seeing this?

**Comment by cmluciano** [comment]
that is quite odd. Is it possible that your k8s-master is also a node?

**Comment by Matty9191** [comment]
The masters are isolated to their own set of VMs. I currently have 3 etcd servers, 3 servers in the control plane and 5 workers. Each host has a unique IP and the POD CIDR and Service CIDR are unique /16s. Glad to provide any additional information. This is an odd one.

**Comment by narasi** [comment]
@cmluciano I can confirm that our master is not a node and it is its own VM too. We just hit upon this problem on 1.7.7

**Comment by narasi** [comment]
@jdumars : We had this problem on 1.6.6, now on 1.7.7. Is there any reason why it can be expected to be fixed in 1.8+?

**Comment by antihax** [comment]
Just had this also happen, fresh k8s setup using kubeadm with weave-net. Master is not a node

```
NAME                           READY     STATUS    RESTARTS   AGE       IP          NODE
nsqadmin-5dd78568f7-srbrk      1/1       Running   0          25s       10.40.0.1   worker01
nsqd-f9f9756df-6kpwh           1/1       Running   0          26s       10.38.0.1   worker03
nsqd-f9f9756df-9c97d           1/1       Running   0          25s       10.38.0.0   worker02
nsqd-f9f9756df-rlhnd           1/1       Running   0          25s       10.40.0.2   worker01
nsqlookupd1-7fd8f9b548-sfqx9   1/1       Running   0          24s       10.38.0.2   worker03
nsqlookupd2-56fc87ccb8-h85z6   1/1       Running   0          24s       10.38.0.1   worker02
```

```
NAME       STATUS    ROLES     AGE       VERSION
master     Ready     master    1d        v1.10.2
worker01   Ready     <none>    1d        v1.10.2
worker02   Ready     <none>    1d        v1.10.2
worker03   Ready     <none>    1d        v1.10.2
```

Trying to see if there is any pattern here.

**Comment by antihax** [comment]
Both worker02 and worker03 had the same MAC address for weave net. The following followed by a reboot solved the issue for me.
```
rm /etc/machine-id
rm /var/lib/dbus/machine-id
systemd-machine-id-setup
```

**Comment by feiskyer** [comment]
@narasi Has the issue solved now? I suspect nodes have been configured with overlapped podCIDRs for such cases. The pod CIDR could be checked by `kubectl get node -o jsonpath='{.items[*].spec.podCIDR}'`

**Comment by khenidak** [comment]
Are you running kubelet as a container? There has been issue mounting the correct volumes @jackfrancis can add more information

**Comment by feiskyer** [comment]
let's close it since it's not active for a long time.

/close

**Comment by Prakashrathod28** [comment]
I'm facing same problem. Can I expect any solution?

**Comment by bimmerkiev** [comment]
The same problem here.

> kubectl get node -o jsonpath='{.items[*].spec.podCIDR}'
gave the different podCIDR for each node.
10.45.1.0/24 10.45.7.0/24 10.45.4.0/24 10.45.0.0/24 10.45.5.0/24 10.45.2.0/24 10.45.3.0/24

**Comment by masaeedu** [comment]
@khenidak I'm running the kubelet (and also the container runtime) inside a container, since I'm using the kind.sigs.k8s.io/ project. Could you elaborate on the issues you were referring to? Is a kubelet or container runtime running inside a container not supported?

**Comment by wuestkamp** [comment]
Same situation using vanilla `kubeadm` installation:

```
root@master1:~# kubectl get pod -A -owide | grep 10.156.0.3
kube-system         etcd-master1                      ...  10.156.0.3   master1
kube-system         kube-apiserver-master1            ...  10.156.0.3   master1
kube-system         kube-controller-manager-master1   ...  10.156.0.3   master1
kube-system         kube-proxy-q87t9                  ...  10.156.0.3   master1
kube-system         kube-scheduler-master1            ...  10.156.0.3   master1
kube-system         weave-net-mndfn                   ...  10.156.0.3   master1
root@master1:~# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.0-rc.2", GitCommit:"27bb2a4a0a5cb8330178d19e57fa61fffa895c98", GitTreeState:"clean", BuildDate:"2020-07-21T17:37:17Z", GoVersion:"go1.14.6", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by Sujal2308** [comment]
> Both worker02 and worker03 had the same MAC address for weave net. The following followed by a reboot solved the issue for me.
> 
> ```
> rm /etc/machine-id
> rm /var/lib/dbus/machine-id
> systemd-machine-id-setup
> ```

**Comment by Sujal2308** [comment]
Hello @antihax  I am facing the same issue of getting same IP address in multiple pods placed in a two different nodes.

I ran your solutions as below:

rm /etc/machine-id
rm /var/lib/dbus/machine-id
systemd-machine-id-setup

after that new machine ID has been generated, but issue is same.

Checked the IP MAC address of both nodes by running ip a there is MAC address is different. 

So kindly let me know how to solve the issue.

Thanks,
Sujal
